validator/telegram/telegram.go

```diff
@@ -0,0 +1,173 @@
+)
+
+type Telegram struct {
+	*validator.Base
+}
+
+const (
+	MATCH_TEMPLATE = "^Sig: (.*)$"
+)
+
+var (
+	client      *telegram.Client
+	l           = logrus.WithFields(logrus.Fields{"module": "validator", "validator": "telegram"})
+	re          = regexp.MustCompile(MATCH_TEMPLATE)
+	POST_STRUCT = map[string]string{
+		"default": "🎭 Verifying my Telegram ID @%s for @NextDotID.\nSig: %%SIG_BASE64%%\n\nPowered by Next.ID - Connect All Digital Identities.\n",
+		"en_US":   "🎭 Verifying my Telegram ID @%s for @NextDotID.\nSig: %%SIG_BASE64%%\n\nPowered by Next.ID - Connect All Digital Identities.\n",
+		"zh_CN":   "🎭 正在通过 @NextDotID 验证我的 Telegram 帐号 @%s 。\nSig: %%SIG_BASE64%%\n\n由 Next.ID 支持 - 连接全域数字身份。\n",
+	}
+)
+
+func Init() {
+	initClient()
+	if validator.PlatformFactories == nil {
+		validator.PlatformFactories = make(map[types.Platform]func(*validator.Base) validator.IValidator)
+	}
+
+	validator.PlatformFactories[types.Platforms.Telegram] = func(base *validator.Base) validator.IValidator {
+		telg := Telegram{base}
+		return &telg
+	}
+}
+
+func (telegram *Telegram) GeneratePostPayload() (post map[string]string) {
+	post = make(map[string]string, 0)
+	for lang_code, template := range POST_STRUCT {
+		post[lang_code] = fmt.Sprintf(template, telegram.Identity)
+	}
+
+	return post
+}
+
+func (telegram *Telegram) GenerateSignPayload() (payload string) {
+	telegram.Identity = strings.ToLower(telegram.Identity)
+	payloadStruct := validator.H{
+		"action":     string(telegram.Action),
+		"identity":   telegram.Identity,
+		"platform":   "telegram",
+		"prev":       nil,
+		"created_at": util.TimeToTimestampString(telegram.CreatedAt),
+		"uuid":       telegram.Uuid.String(),
+	}
+	if telegram.Previous != "" {
+		payloadStruct["prev"] = telegram.Previous
+	}
+
+	payloadBytes, err := json.Marshal(payloadStruct)
+	if err != nil {
+		l.Warnf("Error when marshaling struct: %s", err.Error())
+		return ""
+	}
+
+	return string(payloadBytes)
+}
+
+func (telegram *Telegram) Validate() (err error) {
+	initClient()
+	telegram.Identity = strings.ToLower(telegram.Identity)
+	telegram.SignaturePayload = telegram.GenerateSignPayload()
+	// Deletion. No need to fetch the telegram message.
+	if telegram.Action == types.Actions.Delete {
+		return mycrypto.ValidatePersonalSignature(telegram.SignaturePayload, telegram.Signature, telegram.Pubkey)
+	}
+
+	// Id of the direct message has been sent to the bot.
+	messageId, err := strconv.ParseInt(telegram.ProofLocation, 10, 64)
+	if err != nil {
+		return xerrors.Errorf("Error when parsing telegram message ID %s: %s", telegram.ProofLocation, err.Error())
+	}
+
+	if err := client.Run(context.Background(), func(ctx context.Context) error {
+
+		if _, err := client.Auth().Bot(ctx, config.C.Platform.Telegram.BotToken); err != nil {
+			return xerrors.Errorf("Error when authenticating the telegram bot: %v,", err)
+		}
+
+		msgsClass, err := client.API().MessagesGetMessages(ctx, []tg.InputMessageClass{
```

Comment:
Where should user "post" their proof signature? In any chat? In any public group? Or they *must* post it into our specific, bot-joined public group?

Reply:
Ok, let me explain two different approaches here:
**Appr. A:** 
The user should post the proof directly to the bot after starting it. that's the only place where the bot can read a history of messages. see https://core.telegram.org/method/messages.getMessages.
This is the most hassle-free design (which I used here) according to the telegram API docs. 
Requirements:
- user `api_key` and `api_hash`.
- Bot token for authenticated calls.
Limits:
- I couldn't find any.

**Appr. B:**
We use the user API client and authenticate using 2FA and  Telegram OTP codes...
This way users should post their proofs to a predefined public group, then we read the channel message history using the user API client (authenticated calls). 
Requirements:
- user `api_key` and `api_hash`.
- User authentication for authenticated calls.
Limits:
- There is a Telegram group member limit of 200K members.
- Maintaining a Telegram group.
API:
https://core.telegram.org/method/channels.getMessages.

Reply:
When we designed proof post mechanism, we want *any other user* to validate a specific user's proof material *at any time*. 
So yeah, send proof signature directly to the bot in a private session can finish this binding procedure, but not suitable for *other* people to validate later.

This bothers us a lot to be honest, but there still is a chance to achieve: "forward" user's proof post to a public channel.
Since we can't "falsify" a user's signature, this action can be considered as safe.

If you want to improve this, also make sure to take a consideration of `revalidate` logic: bindings will be `Validate()` again every 7 days, by restoring `validator.Base` from database, and do a `telegram.Validate()` again. So please save `proof_location` to the "forwarded" version of post ID (in a public channel)

Reply:
How about running a bot that posts the received proof to a public channel? If not suited, I can just use the telegram user API and get proofs from a public channel.

Reply:
> How about running a bot that posts the received proof to a public channel?

Yeah that's the plan in my mind.

If the bot can read this public channel's message, it is 100% works.

For other users who wants to validate this message later, they can do it manually (by some way, e.g., copy the message manually and validate it in a small script, etc.). But for our "revalidate" process, this is good enough and can be considered 100% done.

Reply:
Unfortunately, the bot can't read a public channel's message! there will be a simple bot that forwards every proof to a public group. Also, for revalidating we should use the private message to the bot as our input.
If this is not what you wanted, we could just use the core API and a public group! (no need for a telegram bot anymore)

Reply:
🤔 Wow, can a public group do this? 
I mean, if public group can reach the `revalidate` goal, this plan is also perfect!

If a user leaves this group later, will his/her message disappear? If not, ship it!

Thank you again for your great work!

Reply:
Actually, I was able to fetch the telegram message using a proof location of `https://t.me/some_public_group/CHAT_ID_DIGITS`. The only requirement is to add the bot to that public group!